helix-core/src/main/java/org/apache/helix/manager/zk/ZkBaseDataAccessor.java

```diff
@@ -166,14 +195,30 @@ public boolean set(String path, T record, int options) {
    */
   @Override
   public boolean set(String path, T record, int expectVersion, int options) {
-    AccessResult result = doSet(path, record, expectVersion, options);
+    AccessResult result = doSet(path, record, expectVersion, options, true);
+    return result._retCode == RetCode.OK;
+  }
+
+  /**
+   * Sync set operation with custom serializer support
+   */
+  public boolean set(String path, Object record, int options, ZkSerializer zkSerializer) {
+    AccessResult result = doSet(path, zkSerializer.serialize(record), -1, options, false);
+    return result._retCode == RetCode.OK;
+  }
+
+  /**
+   * Sync set operation with custom serializer support
+   */
+  public boolean set(String path, Object record, int options, int expectVersion, ZkSerializer zkSerializer) {
+    AccessResult result = doSet(path, zkSerializer.serialize(record), expectVersion, options, false);
     return result._retCode == RetCode.OK;
   }
 
   /**
    * sync set
    */
-  public AccessResult doSet(String path, T record, int expectVersion, int options) {
+  public AccessResult doSet(String path, Object record, int expectVersion, int options, boolean isZnRecord) {
```

Comment:
public AccessResult doSet(HelixZkClient zkClient, String path, Object record, int expectVersion, int options) {

would be a better structure because it allows for more flexibility (vs 2 options when a boolean is used).

Reply:
The method signatures get fixed in the latest commit.

Reply:
After running the mvn test again, I found another issue with this approach: 
The zkCacheBaseDataAccessor's zkClient == null in some cases (depending on which constructor it is using). So some tests start to fail because of NPE, I feel directly using `_zkclient` is risky in zkCacheBaseDataAccessor. A public get method is safer in that case.

Reply:
Thanks for catching it. In that case, I think using a getter is okay.